File: frpc/csharp-peer/Models/FRpc.cs
using System;
using System.Collections.Generic;
using Fomoxa;

namespace Fomoxa.Rpc.Demo.Models
{
    [Network]
    [Codec("rpc")]
    public class FRpcVoid
    {
    }

    [Network]
    [Codec("rpc")]
    public class FRpcError
    {
        [Network("u32")]
        [Codec("rpc")]
        public uint Code { get; set; }

        [Network("string")]
        [Codec("rpc")]
        public string Message { get; set; } = "";
    }

    [Network]
    [Codec("rpc")]
    public class FRpcCredit
    {
        [Network("u32")]
        [Codec("rpc")]
        public uint Items { get; set; }
    }

    [Network]
    [Codec("rpc")]
    public class FRpcCtx
    {
        [Network("u32")]
        [Codec("rpc")]
        public uint DeadlineMs { get; set; }

        [Network("bytes")]
        [Codec("rpc")]
        public byte[] TraceId { get; set; } = Array.Empty<byte>();

        [Network("bytes")]
        [Codec("rpc")]
        public byte[] SpanId { get; set; } = Array.Empty<byte>();

        [Network("string")]
        [Codec("rpc")]
        public string Token { get; set; } = "";

        [Network("string")]
        [Codec("rpc")]
        public string Tenant { get; set; } = "";

        [Network("bytes")]
        [Codec("rpc")]
        public byte[] IdempotencyKey { get; set; } = Array.Empty<byte>();

        [Network("u32")]
        [Codec("rpc")]
        public uint InitialCredit { get; set; }
    }

    [Network]
    [Codec("rpc")]
    public class FRpcReflectRequest
    {
    }

    [Network]
    [Codec("rpc")]
    public class FRpcMethodInfo
    {
        [Network("string")]
        [Codec("rpc")]
        public string Name { get; set; } = "";

        [Network("u32")]
        [Codec("rpc")]
        public uint Shape { get; set; }

        [Network("u32")]
        [Codec("rpc")]
        public uint RequestId { get; set; }

        [Network("u32")]
        [Codec("rpc")]
        public uint CallerItemId { get; set; }

        [Network("u32")]
        [Codec("rpc")]
        public uint ReplyId { get; set; }

        [Network("u32")]
        [Codec("rpc")]
        public uint RetrySafety { get; set; }
    }

    [Network]
    [Codec("rpc")]
    public class FRpcReflectResponse
    {
        [Network("Array<FRpcMethodInfo>")]
        [Codec("rpc")]
        public List<FRpcMethodInfo> Methods { get; set; } = new List<FRpcMethodInfo>();

        [Network("bytes")]
        [Codec("rpc")]
        public byte[] SchemaJson { get; set; } = Array.Empty<byte>();
    }
}

File: frpc/rust-peer/src/demo/src/models/frpc.rs
#[allow(unused_imports)]
use fomoxa_attributes::{codec, network};

#[derive(Debug, Default, Clone)]
#[network]
#[codec(rpc)]
pub struct FRpcVoid {}

#[derive(Debug, Default, Clone)]
#[network]
#[codec(rpc)]
pub struct FRpcError {
    #[network(u32)]
    #[codec(rpc)]
    pub code: u32,
    #[network(string)]
    #[codec(rpc)]
    pub message: String,
}

#[derive(Debug, Default, Clone)]
#[network]
#[codec(rpc)]
pub struct FRpcCredit {
    #[network(u32)]
    #[codec(rpc)]
    pub items: u32,
}

#[derive(Debug, Default, Clone)]
#[network]
#[codec(rpc)]
pub struct FRpcCtx {
    #[network(u32)]
    #[codec(rpc)]
    pub deadline_ms: u32,
    #[network(bytes)]
    #[codec(rpc)]
    pub trace_id: Vec<u8>,
    #[network(bytes)]
    #[codec(rpc)]
    pub span_id: Vec<u8>,
    #[network(string)]
    #[codec(rpc)]
    pub token: String,
    #[network(string)]
    #[codec(rpc)]
    pub tenant: String,
    #[network(bytes)]
    #[codec(rpc)]
    pub idempotency_key: Vec<u8>,
    #[network(u32)]
    #[codec(rpc)]
    pub initial_credit: u32,
}

#[derive(Debug, Default, Clone)]
#[network]
#[codec(rpc)]
pub struct FRpcReflectRequest {}

#[derive(Debug, Default, Clone)]
#[network]
#[codec(rpc)]
pub struct FRpcMethodInfo {
    #[network(string)]
    #[codec(rpc)]
    pub name: String,
    #[network(u32)]
    #[codec(rpc)]
    pub shape: u32,
    #[network(u32)]
    #[codec(rpc)]
    pub request_id: u32,
    #[network(u32)]
    #[codec(rpc)]
    pub caller_item_id: u32,
    #[network(u32)]
    #[codec(rpc)]
    pub reply_id: u32,
    #[network(u32)]
    #[codec(rpc)]
    pub retry_safety: u32,
}

#[derive(Debug, Default, Clone)]
#[network]
#[codec(rpc)]
pub struct FRpcReflectResponse {
    #[network(Array<FRpcMethodInfo>)]
    #[codec(rpc)]
    pub methods: Vec<FRpcMethodInfo>,
    #[network(bytes)]
    #[codec(rpc)]
    pub schema_json: Vec<u8>,
}
